### PR TITLE
Experiment Container Setup

### PR DESCRIPTION
**Pull request by NathHorrigan**
This PR introduces the idea of 'View Controllers' that manage each of our screens and how they interact with experiment data. We have also added the feature of recovering the experiment if the app has been closed, this is done by introducing a 'Loading' screen at the top of the navigation stack.

### CODE REVIEW

File: src/controllers/ExperimentViewController.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import { LoginScreen, RejectionScreen } from '@screens'
+import { navigate } from '@utils/navigation'
+import { store } from '@redux/store'
+import { experimentSelector, moduleSelector } from '@redux/selectors'
+import { updateExperiment } from '@redux/reducers'
+
+export enum ExperimentModuleType {
+  TermsAndConditions = 'terms',
+  Criteria = 'criteria',
+}
```
Comment (mixxorz) [inline src/controllers/ExperimentViewController.ts]: Suggestion: Don't use typescript enums. It's simpler to use string union types.

```ts
export type ExperimentModuleType = 'TERMS' | 'CRITERIA'
```

File: src/controllers/ExperimentViewController.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  /**
+   * Start presenting the experiment to the particpant.
+   */
+  present() {
+    this.saveExperiment()
+    this.render()
+  }
+
+  /**
+   * Called by a module once it's tasks are finished.
```
Comment (mixxorz) [inline src/controllers/ExperimentViewController.ts]: Typo: its.

File: src/controllers/ExperimentViewController.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+  /**
+   * Utility function to easily present the login screen
+   */
+  static presentLogin() {
+    navigate(LoginScreen.screenID, {})
+  }
+
+  /**
+   * Allows you to recover the experiment if app has been quite mid-experiment
```
Comment (mixxorz) [inline src/controllers/ExperimentViewController.ts]: Typo: quit?

File: src/screens/BaseScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import { NavigationProp } from '@react-navigation/native'
+import { palette } from '@utils/theme'
+
+// Generic utility type used for extract screenID to screen componennt
```
Comment (mixxorz) [inline src/screens/BaseScreen.tsx]: Not sure I understand what this type is for based on this comment.

**Review by mixxorz — APPROVED** [review]
Just a few things! Overall looks good

File: src/controllers/ExperimentViewController.ts (hunk 1 above)
Comment (NathHorrigan) [reply src/controllers/ExperimentViewController.ts]: Habit. Thanks!

File: src/screens/BaseScreen.tsx (hunk 4 above)
Comment (NathHorrigan) [reply src/screens/BaseScreen.tsx]: Updated it. Do you think this is better?

```
/*
  The 2 below types allows us to staticly type the ID of a screen (used by the navigator) and the paramaters that
  should be passed to screen using .navigate method (See react-navigation).
 */
type ModuleScreenProps<Params, Props> = Props & {
  route?: {
    key?: string
    name?: string
    params: Params
  }
}

export type ModuleScreen<
  Params = undefined,
  Props = {}
> = React.FunctionComponent<ModuleScreenProps<Params, Props>> & {
  screenID?: string
}
```

File: src/controllers/ExperimentViewController.ts (hunk 3 above)
Comment (NathHorrigan) [reply src/controllers/ExperimentViewController.ts]: Yep, Thanks!

**Pull request merged**